design/authenticators/storing_non_sensitive_info.md

```diff
@@ -0,0 +1,331 @@
+  * [Version Update](#version-update)
+  * [Delivery Plan](#delivery-plan)
+  
+## Introduction
+
+In several authenticators we need to store some information for their
+authentication process. For example, in `authn-oidc` we have 2 variables
+that store this kind of data:
+  - provider-uri
+  - id-token-user-property
+  
+These 2 variables are defined in the authenticator policy and are loaded with
+values before users can authenticate with the authenticator. However, this data
+is not sensitive and was stored in variables just because we didn't think of a 
+better solution. As the data is not sensitive, it makes sense to store them in
+other methods than in variables, which can improve the UX of the authenticator
+configuration and make another simplicity step.
+
+## Solution
+
+The suggested solution is to store this data in the authenticator webservice's
+annotations. 
+
+For example, let's look at a policy that defines an `authn-oidc`
+authenticator:
+
+```
+- !policy
+  id: conjur/authn-oidc/keycloak
+  body:
+  - !webservice
+    annotations:
+      description: Authentication service for Keycloak, based on Open ID Connect.
+
+  - !variable
+    id: provider-uri
+
+  - !variable
+    id: id-token-user-property
+
+  - !group
+    id: users
+    annotations:
+      description: Group of users who can authenticate using the authn-oidc/keycloak authenticator
+
+  - !permit
+    role: !group users
+    privilege: [ read, authenticate ]
+    resource: !webservice
+```
+
+After loading the policy above, the operator will load the variables with values:
+```
+conjur variable values add conjur/authn-oidc/keycloak/provider-uri "some-provider-uri"
+conjur variable values add conjur/authn-oidc/keycloak/id-token-user-property "some-property"
+```
+
+In the suggested approach, the policy will not have variables, and will be defined
+as follows:
+```
+- !policy
+  id: conjur/authn-oidc/keycloak
+  body:
+  - !webservice
+    annotations:
+      description: Authentication service for Keycloak, based on Open ID Connect.
+      provider-uri: some-provider-uri
+      id-token-user-property: some-property
+
+  - !group
+    id: users
+    annotations:
+      description: Group of users who can authenticate using the authn-oidc/keycloak authenticator
+
+  - !permit
+    role: !group users
+    privilege: [ read, authenticate ]
+    resource: !webservice
+```
+
+This improves the UX of the authenticator configuration for the following reasons:
+  - The user doesn't need to load values into each variable so we have fewer steps
+    of authenticator configuration.
+  - The policy shows clearly all the data of the authenticator, comparing to
+    the current policy where the data of the variables is not visible.
+
+### Authenticator modification
+
+What happens if the user configured an authenticator, and then needs to replace
+one of the non-sensitive data objects? In this case, the user will need to reload
+the policy with the `--delete` flag. 
```

Comment:
`--delete` looks much more scary than it actually is. It will only explicitly delete conjur resources when you use the `!delete`action within the policy. Everything else will remain present and nothing will be deleted implicitly. `--replace` is what should be used very carefully.

Also a lot of customers typically only use replace (especially for authenticators) so I think `--replace` will be used in the wild rather than `--delete`

Reply:
thanks Andrew, I'll add more content to the documentation section that explains what `--delete` does. It also replaces any entity that is present in the policy. 

can you elaborate on how customers use replace for authenticators? How does their policy tree look like? In the docs the policy id for an authenticator is `conjur/authn-something/service-id` so if they will load that into root with `--replace` they will remove everything else from `root`. The only way that will work is if they create a policy named `conjur`, a policy named `authn-something` and then a policy named `service-id` and load with `--replace` into it.

Reply:
@orenbm  It really depends on the customers. Some customers do what you mentioned. They have a `conjur` sub branch and load all authenticators as `--replace` into that subranch. Something like:
`conjur policy load --replace conjur authenticators.yml`. While others have one large policy they load on the root level. That one large policy will have all authenticators and apps and whenever an authenticator needs to be changed/added they will update the root policy and load with `--replace`